astropy/modeling/tests/test_quantities_parameters.py

```diff
@@ -0,0 +1,476 @@
+    assert exc.value.args[0] == ("The .value property on parameters should be set to "
+                                 "unitless values, not Quantity objects. To set a "
+                                 "parameter to a quantity simply set the parameter "
+                                 "directly without using .value")
+
+
+def test_parameter_quantity_property():
+    """
+    Test that the quantity property of Parameters behaves as expected
+    """
+
+    # Since parameters have a .value and .unit parameter that return just the
+    # value and unit respectively, we also have a .quantity parameter that
+    # returns a Quantity instance.
+
+    g = Gaussian1D(1 * u.Jy, 1 * u.m, 0.1 * u.m)
+
+    assert_quantity_allclose(g.amplitude.quantity, 1 * u.Jy)
+
+    # Setting a parameter to a quantity changes the value and the default unit
+    g.amplitude.quantity = 5 * u.mJy
+    assert g.amplitude.value == 5
+    assert g.amplitude.unit is u.mJy
+
+    # But we shouldn't be able to set .quantity to a Quantity with
+    # non-equivalent units
+    with pytest.raises(UnitsError) as exc:
+        g.amplitude.quantity = 3 * u.m
+    assert exc.value.args[0] == ("Cannot set parameter units to m since it is not "
+                                 "equivalent with the original units of mJy")
+
+
+def test_parameter_default_units_match():
+
+    # If the unit and default quantity units are different, raise an error
+    with pytest.raises(ParameterDefinitionError) as exc:
+        class TestC(Fittable1DModel):
+            a = Parameter(default=1.0 * u.m, unit=u.Jy)
+    assert exc.value.args[0] == ("parameter default 1.0 m does not have units "
+                                 "equivalent to the required unit Jy")
+
+
+
+@pytest.mark.parametrize(('unit', 'default'), ((u.m, 1.0), (None, 1 * u.m)))
+def test_parameter_defaults(unit, default):
+    """
+    Test that default quantities are correctly taken into account
+    """
+
+    class TestModel(BaseTestModel):
+        a = Parameter(default=default, unit=unit)
+
+    # TODO: decide whether the default property should return a value or
+    # a quantity?
+
+    # The default unit and value should be set on the class
+    assert TestModel.a.unit == u.m
+    assert TestModel.a.default == 1.0
+
+    # Check that the default unit and value are also set on a class instance
+    m = TestModel()
+    assert m.a.unit == u.m
+    assert m.a.default == m.a.value == 1.0
+
+    # If the parameter is set to a different value, the default is still the
+    # internal default
+    m = TestModel(2.0 * u.m)
+    assert m.a.unit == u.m
+    assert m.a.value == 2.0
+    assert m.a.default == 1.0
+
+    # Instantiate with a different, but compatible unit
+    m = TestModel(2.0 * u.pc)
+    assert m.a.unit == u.pc
+    assert m.a.value == 2.0
+    # The default is still in the original units
+    # TODO: but how do we know what those units are if we don't return a
+    # quantity?
+    assert m.a.default == 1.0
+
```

Comment:
Is the current definition of `Parameter.unit` ambiguous in some cases? I wonder if in addition to `Parameter.unit` there's a need for `Parameter.default_unit` to match `Parameter.default` (which is in reality `Parameter.default_value`). 
Alternatively if `Parameter.default` always returned a Quantity, then `Parameter.unit` could be the  unit for the current `Parameter.value` and NOT the default unit as it is implied now. Then the question is what to do when there's no default value. Is there a valueless quantity (other than just unit)?